### PR TITLE
Add upgrade step on pipeline import

### PR DESCRIPTION
**Pull request by elfenheart**
JIRAs:
- https://issues.cask.co/browse/CDAP-10670 - Fixing Plugin scope between USER vs SYSTEM
- https://issues.cask.co/browse/CDAP-11051 - Missing plugin would ask the user to download from market
- https://issues.cask.co/browse/CDAP-11956 - Do a specific version comparison in addition to check version is in range for pipeline artifact
- https://issues.cask.co/browse/CDAP-11547 - Missing plugin would ask the user to download from market
- https://issues.cask.co/browse/CDAP-9224 - this error message should not happen anymore

## Approach
When you import a pipeline, there is a modal that would appear if there is some problem:
1. Pipeline artifact version/version range does not include current CDAP version
2. Newer plugin artifact version exist
3. Plugin does not exist
4. Plugin exist in different scope

### CODE REVIEW

File: cdap-ui/app/hydrator/controllers/create/toppanel-ctrl.js [hunk 1]
```diff
@@ -872,6 +807,5 @@ class HydratorPlusPlusTopPanelCtrl {
   }
 }
 
-HydratorPlusPlusTopPanelCtrl.$inject = ['$stateParams', 'HydratorPlusPlusConfigStore', 'HydratorPlusPlusConfigActions', '$uibModal', 'HydratorPlusPlusConsoleActions', 'DAGPlusPlusNodesActionsFactory', 'GLOBALS', 'myHelpers', 'HydratorPlusPlusConsoleStore', 'myPipelineExportModalService', '$timeout', '$scope', 'HydratorPlusPlusPreviewStore', 'HydratorPlusPlusPreviewActions', '$interval', 'myPipelineApi', '$state', 'MyCDAPDataSource', 'myAlertOnValium', 'MY_CONFIG', 'PREVIEWSTORE_ACTIONS', '$q', 'NonStorePipelineErrorFactory', 'rArtifacts', '$window', 'LogViewerStore', 'LOGVIEWERSTORE_ACTIONS', 'myPreviewLogsApi', 'DAGPlusPlusNodesStore', 'myPreferenceApi', 'HydratorPlusPlusHydratorService', '$rootScope', 'uuid'];
```
Comment (tonybach) [inline cdap-ui/app/hydrator/controllers/create/toppanel-ctrl.js:875]: We don't need this anymore?

File: cdap-ui/app/hydrator/services/hydrator-upgrade-service.js [hunk 2]
```diff
@@ -0,0 +1,231 @@
...
+  }
+
+  _checkVersionIsInRange(range, version) {
+    if (!range || !version) { return false; }
+
+    if (['[', '('].indexOf(range[0]) !== -1) {
+      const supportedVersion = new window.CaskCommon.Version(version);
+      const versionRange = new window.CaskCommon.VersionRange(range);
+
+      if (versionRange.versionIsInRange(supportedVersion)) {
```
Comment (tonybach) [inline cdap-ui/app/hydrator/services/hydrator-upgrade-service.js]: can simplify this to `return versionRange.versionIsInRange(supportedVersion)`

File: cdap-ui/app/hydrator/services/hydrator-upgrade-service.js [hunk 3]
```diff
@@ -0,0 +1,231 @@
...
+        return true;
+      } else {
+        return false;
+      }
+    }
+
+    // Check equality if range is just a single version
+    if (range !== version) {
```
Comment (tonybach) [inline cdap-ui/app/hydrator/services/hydrator-upgrade-service.js]: `return range === version` here?

**Review by tonybach — COMMENTED** [review]
Couple code comments

File: cdap-ui/app/hydrator/controllers/create/toppanel-ctrl.js (hunk 1 above)
Comment (elfenheart) [reply cdap-ui/app/hydrator/controllers/create/toppanel-ctrl.js:875]: nope, we have `ngInject`

**Comment by tonybach** [comment]
Code LGTM 👍

File: cdap-ui/app/hydrator/services/hydrator-upgrade-service.js [hunk 4]
```diff
@@ -0,0 +1,254 @@
...
+            HydratorPlusPlusConfigStore.setState(HydratorPlusPlusConfigStore.getDefaults());
+            $state.go('hydrator.create', { data: rPipelineConfig });
+          } else {
+            this.loading = false;
+          }
+        };
+
+        checkStages();
+
+        let sub = HydratorPlusPlusLeftPanelStore.subscribe(checkStages);
```
Comment (ajainarayanan) [inline cdap-ui/app/hydrator/services/hydrator-upgrade-service.js:307]: Do we need this subscription? In `checkStages` we don't use left panel store anyways.

**Review by ajainarayanan — COMMENTED** [review]
Just one minor comment. Otherwise 👍 LGTM

File: cdap-ui/app/hydrator/services/hydrator-upgrade-service.js (hunk 4 above)
Comment (elfenheart) [reply cdap-ui/app/hydrator/services/hydrator-upgrade-service.js:307]: Yes we do, it fetches the list of plugins from the left panel store

File: cdap-ui/app/hydrator/templates/create/pipeline-upgrade-modal.html [hunk 5]
```diff
@@ -36,19 +36,19 @@ <h3 class="text-xs-center">
...
 >
-  <h4>
+  <h4 ng-if="PipelineUpgradeController.missingArtifactStages.length > 0 || PipelineUpgradeController.problematicStages.length > 0 || PipelineUpgradeController.problematicPostRunActions.length > 0 || PipelineUpgradeController.missingArtifactPostRunActions.length > 0">
     Your pipeline cannot be imported because of the following issues:
   </h4>
 
   <div
     class="row import-error-row"
-    ng-if="PipelineUpgradeController.missingArtifactStages.length > 0"
+    ng-if="PipelineUpgradeController.missingArtifactStages.length > 0 || problematicPostRunActions.missingArtifactPostRunActions.length > 0"
```
Comment (tonybach) [inline cdap-ui/app/hydrator/templates/create/pipeline-upgrade-modal.html]: I think this should be `PipelineUpgradeController. problematicPostRunActions`?

**Comment by tonybach** [comment]
Code LGTM 👍

**Comment by bdmogal** [comment]
screen shot 2017-08-07 at 4 23 30 pm

Can we mention the missing version here?

**Comment by bdmogal** [comment]
Instead of a user having to click every single one of the plugins to upgrade, can we:
1. Select everything by default
2. Add a ``Select All | Disable All`` toggle button?

image

**Comment by elfenheart** [comment]
@bdmogal not everything has Upgrade button

**Comment by bdmogal** [comment]
LGTM. There is a small improvement about the same plugin showing up multiple times in the missing list. Let's address that as a separate jira.

**Comment by bdmogal** [comment]
Filed https://issues.cask.co/browse/CDAP-12271 for the last issue.

**Pull request merged**